cmd/exo/state.go

```diff
@@ -0,0 +1,192 @@
+
+		var newState map[string]interface{}
+		if err := json.NewDecoder(os.Stdin).Decode(&newState); err != nil {
+			return fmt.Errorf("reading state from stdin: %v", err)
+		}
+
+		workspace := requireWorkspace(ctx, cl)
+		_, err := workspace.SetComponentState(ctx, &api.SetComponentStateInput{
+			Ref:   componentRef,
+			State: jsonutil.MustMarshalString(newState),
+		})
+		return err
+	},
+}
+
+var stateClearCmd = &cobra.Command{
+	Use:    "clear <component>",
+	Short:  "Clear component state",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		workspace := requireWorkspace(ctx, cl)
+		_, err := workspace.SetComponentState(ctx, &api.SetComponentStateInput{
+			Ref:   componentRef,
+			State: "{}",
+		})
+		return err
+	},
+}
+
+var stateEditCmd = &cobra.Command{
+	Use:    "edit <component>",
+	Short:  "Edit component state",
+	Long:   "Edit component state using your preferred editor.",
+	Hidden: true,
+	Args:   cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		componentRef := args[0]
+		ctx := newContext()
+		checkOrEnsureServer()
+		cl := newClient()
+
+		workspace := requireWorkspace(ctx, cl)
+		output, err := workspace.GetComponentState(ctx, &api.GetComponentStateInput{Ref: componentRef})
+		if err != nil {
+			return err
+		}
+
+		tmpfile, err := ioutil.TempFile("", "state.*.json")
+		if err != nil {
+			return fmt.Errorf("creating temporary file: %w", err)
+		}
+		defer os.Remove(tmpfile.Name())
+
+		if err := jsonutil.PrettyPrintJSONString(tmpfile, output.State); err != nil {
+			return fmt.Errorf("writing to temporary file: %w", err)
+		}
+		if err := tmpfile.Close(); err != nil {
+			return fmt.Errorf("closing temporary file: %w", err)
+		}
+
+		// TODO: Read from EDITOR, or guess from common editors.
+		editor := os.Getenv("EDITOR")
+		if editor == "" {
+			wd, err := os.Getwd()
+			if err != nil {
+				return fmt.Errorf("getting working directory: %w", err)
+			}
+			pathVar, _ := os.LookupEnv("PATH")
+
+			for _, candidateEditor := range []string{
+				"code",
+				"vim",
+				"nano",
+				"vi",
+				"emacs",
+				"ee",
+			} {
+				found, err := which.Query{
+					WorkingDirectory: wd,
+					PathVariable:     pathVar,
+					Program:          candidateEditor,
+				}.Run()
+				if err != nil && !strings.Contains(err.Error(), "not found") {
+					return fmt.Errorf("looking up candidate editor %q: %w", candidateEditor, err)
+				}
+				if found != "" {
+					editor = found
+					break
+				}
+			}
+
+			if editor == "" {
+				return errors.New("No editor available")
+			}
+		}
+
+		edit := exec.Command(editor, tmpfile.Name())
+		edit.Stdin = os.Stdin
+		edit.Stdout = os.Stdout
+		edit.Stderr = os.Stderr
+		if err := edit.Run(); err != nil {
+			return fmt.Errorf("editing state file: %w", err)
+		}
+
+		newState, err := os.ReadFile(tmpfile.Name())
```

Comment:
Minor but I'd argue that we should only update the state if the temporary file was saved. Otherwise we leave people no way of exiting without doing an update.

In an ideal world, the temp file would act as a two way binding to the component's state.

Reply:
Good call. I'll update the code to do this.